### PR TITLE
Adding saml migration command

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
I have splited the auth migration to two commads because the parameters changes
from one command to another.

Probably is necesary too to review the texts. The new command format is:

  - `platform user migrate_auth_to_saml [fromAuth] [usersFile.json] [--dryRun] [--auto]`
  - `platform user migrate_auth_to_ldap [fromAuth] [matchField] [--dryRun] [--force]`

Other option would be something like `platform user migrate_auth [ldap|saml]` being ldap and saml subcommands of migrate_auth.

#### Ticket Link
[XYZ-31](https://mattermost.atlassian.net/browse/XYZ-31)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [X] Has enterprise changes (https://github.com/mattermost/enterprise/pull/261)
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: cmd/platform/user.go [hunk 1]
```diff
@@ -86,26 +89,58 @@ var deleteAllUsersCmd = &cobra.Command{
 	RunE:    deleteAllUsersCommandF,
 }
 
-var migrateAuthCmd = &cobra.Command{
-	Use:   "migrate_auth [from_auth] [to_auth] [match_field]",
-	Short: "Mass migrate user accounts authentication type",
-	Long: `Migrates accounts from one authentication provider to another. For example, you can upgrade your authentication provider from email to ldap.
+var migrateAuthToSamlCmd = &cobra.Command{
```
Comment (enahum) [inline cmd/platform/user.go]: why the change of name here? were you planning to add this as another option?

Comment (jespino) [reply cmd/platform/user.go]: Yes, I have created 2 differente commands, because the previous one is quite different in the expected parameters. I haven't a strong opinion here about how must be the commands, but based on how cobra works, have more sense 2 have 2 different commands which receive differente parameters. In the PR description I include another proposal (based on migrate_auth subcommands). I can merge all together in the old migrateAuth command, but I think is more confusing and difficult to explain the command in the help.

Comment (jespino) [reply cmd/platform/user.go]: Actually I expect some kind of feedback from @jasonblais about this new interface.

File: cmd/platform/user.go [hunk 2]
```diff
@@ -86,26 +89,58 @@ var deleteAllUsersCmd = &cobra.Command{
 	RunE:    deleteAllUsersCommandF,
 }
 
-var migrateAuthCmd = &cobra.Command{
-	Use:   "migrate_auth [from_auth] [to_auth] [match_field]",
```
Comment (jwilander) [inline cmd/platform/user.go]: Changing this command could potentially break someone's script. Could we keep the command the same and just document the different options for the different `to_auth` types?

Comment (jespino) [reply cmd/platform/user.go]: That's a good point. I'm not sure if this is the kind of command that de people use periodically, but anyway, there exists the possibility that we are breaking something.

I think the resulting "help" information will be more confusing, but if we don't want to break anything, is probably the best option.

Comment (jespino) [reply cmd/platform/user.go]: Ok, finally I merged again the commands and rewrite the help of the commands. Now I think the result is clear and the API hasn't changed (except for the extra --dryRun param in the ldap command).

File: cmd/platform/user.go [hunk 3]
```diff
@@ -138,7 +157,69 @@ func init() {
...
+	migrateAuthCmd.SetUsageTemplate(`Usage:
+  platform user migrate_auth [from_auth] [to_auth] [migration-options] [flags]
+
+Examples:
+{{.Example}}
+
+Arguments:
+  from_auth:
+    The authentication service to migrate users accounts from.
+    Supported options: email, gitlab, ldap.
```
Comment (jasonblais) [inline cmd/platform/user.go]: I think `saml` is supported as well

File: i18n/en.json [hunk 4]
```diff
@@ -4174,6 +4174,18 @@
     "id": "ent.migration.migratetoldap.user_not_found",
     "translation": "Unable to find user on AD/LDAP server: "
   },
+  {
+    "id": "ent.migration.migratetosaml.user_not_found_in_users_mapping_file",
+    "translation": "User not found in the users file."
+  },
+  {
+    "id": "ent.migration.migratetosaml.username_already_used_by_other_user",
+    "translation": "Username already used by other mattermost user."
```
[the remark this answers is not in the document]
Comment (jespino) [reply i18n/en.json]: Done

File: i18n/en.json [hunk 5]
```diff
@@ -4174,6 +4174,18 @@
...
+  },
+  {
+    "id": "ent.migration.migratetosaml.email_already_used_by_other_user",
+    "translation": "Email already used by other SAML user."
```
[the remark this answers is not in the document]
Comment (jespino) [reply i18n/en.json]: Done

File: cmd/platform/user.go [hunk 6]
```diff
@@ -87,25 +90,41 @@ var deleteAllUsersCmd = &cobra.Command{
...
+		}
 
-Will display any accounts that are not migrated successfully.`,
-	Example: "  user migrate_auth email ladp email",
-	RunE:    migrateAuthCmdF,
+		autoFlag, _ := cmd.Flags().GetBool("auto")
+
+		if toAuth == "saml" && autoFlag {
+			if len(args) != 2 {
+				return errors.New("Saml migration requires two argument when use --auto flag. See help text for details.")
```
[the remark this answers is not in the document]
Comment (jespino) [reply cmd/platform/user.go]: Done

File: cmd/platform/user.go [hunk 7]
```diff
@@ -87,25 +90,41 @@ var deleteAllUsersCmd = &cobra.Command{
...
+			}
+		}
+
+		if toAuth == "saml" && !autoFlag {
+			if len(args) != 3 {
+				return errors.New("Saml migration requires three arguments when not use --auto flag. See help text for details.")
```
[the remark this answers is not in the document]
Comment (jespino) [reply cmd/platform/user.go]: Done

File: cmd/platform/user.go [hunk 8]
```diff
@@ -138,7 +157,69 @@ func init() {
 
 	deleteAllUsersCmd.Flags().Bool("confirm", false, "Confirm you really want to delete the user and a DB backup has been performed.")
 
-	migrateAuthCmd.Flags().Bool("force", false, "Force the migration to occour even if there are duplicates on the LDAP server. Duplicates will not be migrated.")
+	migrateAuthCmd.Flags().Bool("force", false, "Force the migration to occour even if there are duplicates on the LDAP server. Duplicates will not be migrated. (ldap only)")
```
[the remark this answers is not in the document]
Comment (jespino) [reply cmd/platform/user.go]: Done

File: cmd/platform/user.go [hunk 9]
```diff
@@ -138,7 +157,69 @@ func init() {
...
+	migrateAuthCmd.Flags().Bool("auto", false, "Auto migrate all users assuming that the SAML service usernames and emails are identical to the mattermost usernames and emails. (saml only)")
```
[the remark this answers is not in the document]
Comment (jespino) [reply cmd/platform/user.go]: Done

File: cmd/platform/user.go (hunk 3 above)
Comment (jespino) [reply cmd/platform/user.go]: You are right, and done :)

**Comment by jespino** [comment]
>  I cannot see the enterprise changes (don't have access), but I assume those are okay
Yes, the enterprise changes are in the enterprise repository, and is being reviewed there.

>  I agree with Joram not to cause breaking changes. The updated implementation looks good where SAML migration happens via `platform user migrate_auth {from_service} saml
Yes, me too, I changed the command again to have exactly that design, I have to get out of the cobra's (CLI generation library) way to do the things, but It's fine, the new approach is supported by cobra and I think works well.

> What are your thoughts on Corey's comment from today where the --auto command might tempt admins to just run it, and then hit issues? Wondering if it's safer to force them to use the more manual process for migrating users
I had the same concern when I wrote the code, but we have put a pretty clear warning message, which is: `You are about to perform an automatic \"" + fromAuth + " to saml\" migration, this must be done only if you are sure your current mattermost users using " + fromAuth + " auth have the same username and email in your SAML service, otherwise you must provide the users and emails from your Saml Service using the \"users file\" instead of \"--auto\" option. Do you want to proceed with auto migration anyway? (YES/NO):` we can start the message with a "WARNING:" text, to be more explicit.

> Do we / should we test JSON exports from Onelogin/Okta/ADFS work as expected?
Maybe, but the way SAML integration works, it doesn't matter too much. We have the AuthService set to saml, and the AuthData set to the email, if somebody comes from the configured saml service, and the user data is authorized with that emal, then, the user is in. We can test, but I'm don't see exactly what can goes wrong, strictly related to the Onelogin/Okta/ADFS services. But maybe I'm not seeing something. Probably @enahum has a stronger and better founded opinion here

> How easy is it to extend to other accepted formats like CSV?
Actually is pretty easy. We can check for file extension, and then try to parse the JSON or the CSV, or any other format.

**Comment by jespino** [comment]
The build is failing because is using other enterprise branch (I don't know exactly why).

**Comment by enahum** [comment]
I’ve never done an export with any of those services so I don’t have a clue on what and how it is exported asumming an export is possible.

I think we should document the file format that is expected and people using the tool will need to provide it as requested

**Comment by jespino** [comment]
An example with OneLogin would be something like this (I have to test it, but I haven't a OneLogin account now):

```
#!/usr/bin/env python

from onelogin.api.client import OneLoginClient
import json

client_id = input("Client id: ")
client_secret = input("Secret: ")
region = input("Region (us, eu): ")

client = OneLoginClient(client_id, client_secret, region)

mapping = {}
for user in client.get_users():
    mapping[user.email] = user.username

print(json.dumps(mapping))
```

And you only need to have python installed and install the onelogin python sdk with the command `pip install onelogin`.

**Comment by jespino** [comment]
And for Okta, would be something like this:
```
#!/usr/bin/env python
from okta import UsersClient
import json

base_url = input("Base url (example: https://example.okta.com): ")
api_token = input("API Token: ")
                  
usersClient = UsersClient(base_url, api_token)

users = usersClient.get_paged_users(limit=25)
       
mapping = {}

for user in users.result:
    mapping[user.email] = user.login
                         
while not users.is_last_page():
    users = usersClient.get_paged_users(url=users.next_url)
    for user in users.result:
        mapping[user.email] = user.login

print(json.dumps(mapping))
```

**Comment by jespino** [comment]
I think we can test that scripts working, but I think is not needed the whole integration (users export + user migration), because the user migration is pretty independent from the service itself.

**Comment by jasonblais** [comment]
Thanks @jespino - all your notes look good. We can add (and test) the sample scripts when working on the documentation. Good to know it'll be easy to extend the import format if a customer needs CSV for instance.

For the `--auto` command text, that's good then. I'm 0/5 on the use of `WARNING`, is that common for CLI? Either way, proposing a slight tweak to the prompt text:

```
WARNING: You are about to perform an automatic \"" + fromAuth + " to saml\" migration. This must be done only if you are sure your current Mattermost users using " + fromAuth + " auth have the same username and email in your SAML service. Otherwise, you must provide the usernames and emails from your SAML Service using the \"users file\" instead of \"--auto\" option. Do you want to proceed with automatic migration anyway? (YES/NO):
```

**Comment by jespino** [comment]
About the `--auto`, the most dangerous and undoable command in the CLI is `platform reset` and first ask you: `Have you performed a database backup? (YES/NO): `, and then ask you `Are you sure you want to delete everything? All data will be permanently deleted? (YES/NO): ` if you answer something different thant YES (all capital letters) the operation will be aborted.

So, the message without the "WARNING" would be the consistent way to go, but we can make it consistent changing the reset command and any other dangerous command.

I going to start updating the current text (without the WARNING).

**Comment by jasonblais** [comment]
@jespino Sounds good, let's not add the warning text. Sorry, I noticed a grammar mistake in my proposed text, fixed below. Also propose we have the question on a separate line so that it's easier to see.

```
You are about to perform an automatic \"" + fromAuth + " to saml\" migration. This must only be done if your current Mattermost users with " + fromAuth + " auth have the same username and email in your SAML service. Otherwise, provide the usernames and emails from your SAML Service using the \"users file\" without the \"--auto\" option. 

Do you want to proceed with automatic migration anyway? (YES/NO):
```

**Comment by jespino** [comment]
Done, used the last version of the text, and with one line separation. Now if @jwilander and @enahum are fine with the changes we are done :)

**Comment by jwilander** [comment]
With a rebase we're good to go 👍

**Comment by jespino** [comment]
Rebased

**Comment by jasonblais** [comment]
@jespino You can merge this PR 👍
